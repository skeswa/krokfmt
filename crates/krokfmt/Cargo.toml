[package]
name = "krokfmt"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
authors.workspace = true
description = "A highly opinionated TypeScript code formatter"
license.workspace = true
repository.workspace = true
readme = "../../README.md"
keywords = ["typescript", "formatter", "code-formatter", "tsx", "jsx"]
categories = ["development-tools", "command-line-utilities"]

[[bin]]
name = "krokfmt"
path = "src/main.rs"

[dependencies]
# NOTE: we have to resort to some tomfoolery to get biome deps working. This is
# due mostly to the unwillingness of the maintainers to ensure their crates are
# stable.
#
# I yanked these version pins from https://github.com/brioche-dev/brioche/pull/184.
biome_console = "=0.5.7"
biome_deserialize = "=0.5.7"
biome_deserialize_macros = "=0.5.7"
biome_diagnostics = "=0.5.7"
biome_diagnostics_categories = "=0.5.7"
biome_diagnostics_macros = "=0.5.7"
biome_formatter = "=0.5.7"
biome_js_factory = "=0.5.7"
biome_js_formatter = "=0.5.7"
biome_js_parser = "=0.5.7"
biome_js_syntax = "=0.5.7"
biome_json_factory = "=0.5.7"
biome_json_parser = "=0.5.7"
biome_json_syntax = "=0.5.7"
biome_markup = "=0.5.7"
biome_parser = "=0.5.7"
biome_rowan = "=0.5.7"
biome_text_edit = "=0.5.7"
biome_text_size = "=0.5.7"
biome_unicode_table = "=0.5.7"
clap = { version = "4.5", features = ["derive"] }
swc_atoms = "^7.0.0"
swc_common = { version = "^14.0.0", features = ["tty-emitter"] }
swc_ecma_ast = "^15.0.0"
swc_ecma_codegen = "^16.0.0"
swc_ecma_parser = "^22.0.3"
swc_ecma_visit = "^14.0.0"
anyhow = { workspace = true }
glob = "0.3"
rayon = "1.8"
colored = "2.1"

[dev-dependencies]
tempfile = "3.9"
insta = "1.34"
pretty_assertions = "1.4"
criterion = { version = "0.5", features = ["html_reports"] }

[[bench]]
name = "formatting_bench"
harness = false

[[bench]]
name = "real_world_bench"
harness = false