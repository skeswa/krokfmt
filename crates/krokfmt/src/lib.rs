pub mod biome_formatter;
pub mod codegen;
pub mod comment_classifier;
pub mod comment_extractor;
pub mod comment_formatter;
pub mod comment_reinserter;
pub mod file_handler;
pub mod organizer;
pub mod parser;
pub mod selective_comment_handler;
pub mod semantic_hash;
pub mod transformer;

use anyhow::{Context, Result};
use std::path::Path;

/// Simple heuristic to detect JSX content in source code.
/// Looks for common JSX patterns like <Component> or JSX expressions.
fn contains_jsx(source: &str) -> bool {
    // Look for JSX element patterns: < followed by uppercase letter or lowercase HTML tag
    // This is a simple heuristic that covers most cases
    source.contains("</") || source.contains("/>") || 
    source.contains("React.") || source.contains("jsx") ||
    // Check for common JSX patterns
    source.chars().zip(source.chars().skip(1)).any(|(c1, c2)| {
        c1 == '<' && (c2.is_ascii_uppercase() || c2.is_ascii_lowercase())
    })
}

/// Format TypeScript/TSX code with krokfmt's opinionated rules.
///
/// This is the main entry point for programmatic use of krokfmt.
/// It applies the full formatting pipeline: parsing, organizing, and final formatting.
pub fn format_typescript(source: &str, filename: &str) -> Result<String> {
    // Auto-detect JSX content and use appropriate extension
    let has_jsx = contains_jsx(source);
    let effective_filename =
        if !filename.ends_with(".tsx") && !filename.ends_with(".jsx") && has_jsx {
            // If the filename doesn't already indicate JSX/TSX and we detected JSX, use .tsx
            "input.tsx".to_string()
        } else if filename.ends_with(".ts") && has_jsx {
            // If it's explicitly .ts but contains JSX, convert to .tsx
            filename.replace(".ts", ".tsx")
        } else {
            filename.to_string()
        };

    // Parse the TypeScript code
    let parser = parser::TypeScriptParser::new();
    let source_map = parser.source_map.clone();
    let comments = parser.comments.clone();
    let module = parser
        .parse(source, &effective_filename)
        .context("Failed to parse TypeScript code")?;

    // Organize the code structure with selective comment preservation
    let formatter = comment_formatter::CommentFormatter::new(source_map, comments);
    let organized_content = formatter
        .format(module, source)
        .context("Failed to organize code")?;

    // Apply final formatting with Biome
    let biome_formatter = biome_formatter::BiomeFormatter::new();
    let formatted_content = biome_formatter
        .format(&organized_content, Path::new(&effective_filename))
        .context("Failed to format with Biome")?;

    Ok(formatted_content)
}
