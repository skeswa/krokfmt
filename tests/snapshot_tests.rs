use insta::assert_snapshot;
use krokfmt::{codegen::CodeGenerator, formatter::KrokFormatter, parser::TypeScriptParser};
use std::fs;

fn format_code(input: &str) -> String {
    let parser = TypeScriptParser::new();
    let source_map = parser.source_map.clone();
    let filename = if input.contains("<") && input.contains(">") {
        "test.tsx"
    } else {
        "test.ts"
    };
    let module = parser.parse(input, filename).unwrap();
    let formatted = KrokFormatter::new().format(module).unwrap();
    let generator = CodeGenerator::new(source_map);
    generator.generate(&formatted).unwrap()
}

fn test_fixture(fixture_path: &str) {
    let input_path = format!("tests/fixtures/{}.input.ts", fixture_path);
    let input = fs::read_to_string(&input_path)
        .unwrap_or_else(|_| panic!("Failed to read fixture: {}", input_path));
    
    let output = format_code(&input);
    
    // Use the fixture path as the snapshot name
    assert_snapshot!(fixture_path, output);
}

// FR1: Import/Export Organization Tests

#[test]
fn test_fr1_1_default_imports() {
    test_fixture("fr1/1_1_default_imports");
}

#[test]
fn test_fr1_1_named_imports() {
    test_fixture("fr1/1_1_named_imports");
}

#[test]
fn test_fr1_1_namespace_imports() {
    test_fixture("fr1/1_1_namespace_imports");
}

#[test]
fn test_fr1_1_side_effect_imports() {
    test_fixture("fr1/1_1_side_effect_imports");
}

#[test]
fn test_fr1_1_type_imports() {
    test_fixture("fr1/1_1_type_imports");
}

#[test]
fn test_fr1_1_mixed_imports() {
    test_fixture("fr1/1_1_mixed_imports");
}

#[test]
fn test_fr1_1_import_aliases() {
    test_fixture("fr1/1_1_import_aliases");
}

#[test]
fn test_fr1_2_categorization() {
    test_fixture("fr1/1_2_categorization");
}

#[test]
fn test_fr1_3_sorting() {
    test_fixture("fr1/1_3_sorting");
}

#[test]
fn test_fr1_4_positioning() {
    test_fixture("fr1/1_4_positioning");
}

#[test]
fn test_fr1_5_group_separation() {
    test_fixture("fr1/1_5_group_separation");
}

#[test]
fn test_fr1_6_syntax_preservation() {
    test_fixture("fr1/1_6_syntax_preservation");
}

// FR2: Member Visibility Ordering Tests

#[test]
fn test_fr2_1_export_detection() {
    test_fixture("fr2/2_1_export_detection");
}

#[test]
fn test_fr2_2_export_prioritization() {
    test_fixture("fr2/2_2_export_prioritization");
}

#[test]
fn test_fr2_3_dependency_preservation() {
    test_fixture("fr2/2_3_dependency_preservation");
}

#[test]
fn test_fr2_4_intelligent_grouping() {
    test_fixture("fr2/2_4_intelligent_grouping");
}

// FR3: Alphabetical Sorting Tests

#[test]
fn test_fr3_1_function_arguments() {
    test_fixture("fr3/3_1_function_arguments");
}

#[test]
fn test_fr3_2_object_properties() {
    test_fixture("fr3/3_2_object_properties");
}

#[test]
fn test_fr3_3_class_members() {
    test_fixture("fr3/3_3_class_members");
}

#[test]
fn test_fr3_4_type_members() {
    test_fixture("fr3/3_4_type_members");
}

#[test]
fn test_fr3_5_enum_members() {
    test_fixture("fr3/3_5_enum_members");
}

#[test]
fn test_fr3_6_jsx_properties() {
    test_fixture("fr3/3_6_jsx_properties");
}