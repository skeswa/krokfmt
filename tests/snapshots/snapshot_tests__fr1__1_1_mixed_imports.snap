---
source: tests/snapshot_tests.rs
expression: output
---
import axios, { AxiosError } from 'axios';
import React, { useState, useEffect } from 'react';

import utils, { helper, type HelperType } from '../utils/helpers';

export function useData() {
    const [data, setData] = useState(null);
    useEffect(()=>{
        axios.get('/api/data').then((res)=>setData(res.data));
    }, []);
    return data;
}
